Update `IGroupsClient` to support paging of groups and projects

belcher-rok opened the pull request:
Added ability to enumerate groups, subgroups, and projects by page number. Each successful call returns a collection of items in addition to the total number of items.

All paging methods start with "Page", all are async, and they use `GroupId` to id the target group where applicable.

Also
- Added ability to query top-level groups (see `GroupQuery.TopLevelOnly`).
- Added ability to query descendant subgroups (see `SubgroupQuery.IncludeDescendants`).
- Added ability for `IHttpRequestor` to access the response headers if necessary.
- Added equality operators to `GroupId` and `ProjectId`.
- Added overloads taking a `GroupId` as long as doing so would not result in ambiguous calls for existing methods.
- All changes should be backward compatible, at least that is the intention.

belcher-rok committed:
Add Page support to IGroupsClient

belcher-rok committed:
Add unit tests for mock GroupClient

belcher-rok committed:
Add integration tests

belcher-rok commented on NGitLab/Models/GroupId.cs:
```diff
@@ -13,24 +24,89 @@ namespace NGitLab.Models;
         _path = path ?? throw new ArgumentNullException(nameof(path));
     }
 
-    public GroupId(Group group)
```
Removed this c'tor because the resulting `GroupId` is ambiguous. For example,
```C#
Group group = GetGroup(...);
var id1 = new GroupId(group); // calls the removed c`tor
var id2 = new GroupId(group.Id);
var id3 = new GroupId(group.FullPath);

Assert(id1 == id2);  // Ok, both contain `group.Id`
Assert(id1 == id3);  // Fails! group.Id != group.FullPath.
Assert(id2 == id3);  // Fails for same reason, but there no easy way around this other than to not do it.

// Preferred usage:
Assert(id1 == group); // Ok!
Assert(id2 == group); // Ok!
Assert(id3 == group); // Ok!
```

louis-z replied:
It's probably wise to eventually get rid of this ctor, but for now please leave it in and tag is with the `[Obsolete]` attribute, so as not to introduce a breaking change.

belcher-rok committed:
Merge branch 'main' into main

belcher-rok committed:
Fix integration test failures

belcher-rok commented:
@louis-z Do you need anything else from me?

louis-z commented:
> @louis-z Do you need anything else from me?

I haven't had the chance to look at this PR yet. Will let you know...

louis-z commented on NGitLab/IHttpRequestor.cs:
```diff
@@ -12,10 +13,18 @@ public interface IHttpRequestor
+
     Task StreamAsync(string tailAPIUrl, Func<Stream, Task> parser, CancellationToken cancellationToken);
 
+    Task StreamAndHeadersAsync(string tailAPIUrl, Func<Stream, WebHeaderCollection, Task> parser, CancellationToken cancellationToken);
```
Do not expose in the public API any of `System.Net`'s types. We have plans to eventually steer away from the `Web*` classes (e.g. `WebRequest`) completely and use `HttpClient` instead.

Maybe you could expose headers via a `IReadOnlyDictionary<string, string>` instead?

belcher-rok replied:
Switched to `IReadOnlyDictionary<string, IEnumerable<string>>`. I had to create an adaptor class to do it. See `WebHeadersDictionaryAdaptor`.

louis-z commented on NGitLab/IGroupsClient.cs:
```diff
@@ -9,53 +9,183 @@ namespace NGitLab;
-    GitLabCollectionResponse<Project> GetProjectsAsync(int groupId, GroupProjectsQuery query);
+    /// <inheritdoc cref="SearchProjectsAsync"/>
+    /// <param name="id">The group's id</param>
+    GitLabCollectionResponse<Project> GetProjectsAsync(int id, GroupProjectsQuery query);
```
Don't rename `groupId` to `id`, as this is a breaking change.

louis-z commented on NGitLab/IGroupsClient.cs:
```diff
@@ -9,53 +9,183 @@ namespace NGitLab;
+    /// <param name="id">The group's id</param>
+    /// <param name="search">The search criteria.</param>
+    [Obsolete("Use SearchProjectsAsync instead")]
+    IEnumerable<Project> SearchProjects(int id, string search);
```
Don't rename `groupId` to `id`, as this is a breaking change.

louis-z commented on NGitLab/Models/ProjectId.cs:
```diff
@@ -13,24 +24,89 @@ namespace NGitLab.Models;
         _path = path ?? throw new ArgumentNullException(nameof(path));
     }
 
-    public ProjectId(Project project)
```
Please leave this ctor in and simply tag it `[Obsolete]` for now.

louis-z commented:
Hey @belcher-rok, thanks a lot for your contribution! This is great work!

Adding/improving comments was a nice touch too, but when documenting a method, could you please use the 3rd person (*descriptive form*, such as `gets`) rather than the 2nd person (*prescriptive form*, such as `get`)? This is a widespread standard which, unfortunately, we did not always comply with in the past.

Check out
- [The Google doc guidelines](https://developers.google.com/style/reference-verbs)
- [An example from Microsoft doc](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.getattributes?view=net-6.0#definition)

Toa741 commented on NGitLab.Mock.Tests/GroupsMockTests.cs:
```diff
@@ -78,4 +151,192 @@ public void Test_get_subgroups_by_fullpath()
+        Assert.ThrowsAsync<Clients.GitLabBadRequestException>(() => client.Groups.PageSubgroupsAsync(1, new(page: 1, perPage: 0)));
+    }
+
+    private static GitLabServer CreateProjectHierarchy() =>
```
Can it be moved at the top of the file instead? Even though I understand it was to keep it close to the related tests, let's keep static fields together.

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -155,28 +122,66 @@ public IEnumerable<Models.Group> Get(GroupQuery query)
+        if (perPage < PageQuery.MinPerPage)
+        {
+            // Max isn't enforced the same way
+            throw new GitLabBadRequestException("per_page has a value not allowed");
```
```suggestion
            throw new GitLabBadRequestException($"per_page value ({perPage}) is invalid: cannot be lower than MinPerPage ({PageQuery.MinPerPage})");
```

To give more context.

belcher-rok replied:
@Toa741 I don't have a problem making this change, I was just attempting to mimic the [actual GitLab error message](https://gitlab.com/gitlab-org/gitlab/-/blob/master/spec/requests/api/commits_spec.rb#L356). Should I still use a more descriptive message?

Toa741 replied:
Yes, the mock is primarily used for tests and I would be happier with a message that points the problem immediately, rather than trying to find the response in the doc.

Ideally, the API response would be more precise too ^^, `per_page has a value not allowed` is a bit vague

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -155,28 +122,66 @@ public IEnumerable<Models.Group> Get(GroupQuery query)
+            throw new GitLabBadRequestException("per_page has a value not allowed");
+        }
+
+        var all = Get(query?.Query).ToList();
```
```suggestion
        var all = Get(query?.Query).ToArray();
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -155,28 +122,66 @@ public IEnumerable<Models.Group> Get(GroupQuery query)
+        var page = all
+            .Skip((pageNum - 1) * perPage)
+            .Take(perPage)
+            .ToList();
```
```suggestion
            .ToArray();
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -231,6 +240,24 @@ public GitLabCollectionResponse<Models.Project> GetProjectsAsync(int groupId, Gr
+        var perPage = query?.PerPage ?? PageQuery.DefaultPerPage;
+        if (perPage < 1)
+        {
+            throw new GitLabBadRequestException("per_page has a value not allowed");
```
```suggestion
            throw new GitLabBadRequestException("per_page value ({perPage}) is invalid: cannot be inferior to 1");
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -231,6 +240,24 @@ public GitLabCollectionResponse<Models.Project> GetProjectsAsync(int groupId, Gr
+            throw new GitLabBadRequestException("per_page has a value not allowed");
+        }
+
+        var all = SearchProjectsAsync(groupId, query?.Query).ToList();
```
```suggestion
        var all = SearchProjectsAsync(groupId, query?.Query).ToArray();
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -231,6 +240,24 @@ public GitLabCollectionResponse<Models.Project> GetProjectsAsync(int groupId, Gr
+        var page = all
+            .Skip((pageNum - 1) * perPage)
+            .Take(perPage)
+            .ToList();
```
```suggestion
            .ToArray();
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -320,42 +352,49 @@ public GitLabCollectionResponse<Models.Group> GetSubgroupsByIdAsync(int id, Subg
+        var perPage = query?.PerPage ?? PageQuery.DefaultPerPage;
+        if (perPage < 1)
+        {
+            throw new GitLabBadRequestException("per_page has a value not allowed");
```
```suggestion
            throw new GitLabBadRequestException("per_page value ({perPage}) is invalid: cannot be inferior to 1");
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -320,42 +352,49 @@ public GitLabCollectionResponse<Models.Group> GetSubgroupsByIdAsync(int id, Subg
+            throw new GitLabBadRequestException("per_page has a value not allowed");
         }
+
+        var all = GetSubgroupsAsync(groupId, query?.Query).ToList();
```
```suggestion
        var all = GetSubgroupsAsync(groupId, query?.Query).ToArray();
```

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -320,42 +352,49 @@ public GitLabCollectionResponse<Models.Group> GetSubgroupsByIdAsync(int id, Subg
+        var page = all
+            .Skip((pageNum - 1) * perPage)
+            .Take(perPage)
+            .ToList();
```
```suggestion
            .ToArray();
```

Toa741 commented on NGitLab/IGroupsClient.cs:
```diff
@@ -9,53 +9,183 @@ namespace NGitLab;
+    /// The page will be empty after the last page is read.<para/>
+    /// The total maybe null if there are more than 10,000 items. See https://docs.gitlab.com/ee/user/gitlab_com/index.html#pagination-response-headers.
+    /// </remarks>
+    Task<(IReadOnlyCollection<Group> Page, int? Total)> PageAsync(PageQuery<GroupQuery> query, CancellationToken cancellationToken = default);
```
Not a huge fan of exposed Tuples. It looks a bit cumbersome as it's repeated in multiple places. Although it might not happen, if we need to add a property in the future it makes a lot of place to change too. 

Could we create a simple class for it (or a record)? `PaginatedResponse<T>` or something equivalent

belcher-rok replied:
Good idea. Created `PagedResponse<T>` type.

Toa741 commented on NGitLab.Mock/Clients/GroupClient.cs:
```diff
@@ -320,42 +352,49 @@ public GitLabCollectionResponse<Models.Group> GetSubgroupsByIdAsync(int id, Subg
+            .Skip((pageNum - 1) * perPage)
+            .Take(perPage)
+            .ToList();
+        return (page, all.Count > 10000 ? null : all.Count);
```
Can the 10000 limit be added as a const?

Toa741 commented on NGitLab/Impl/GroupsClient.cs:
```diff
@@ -36,61 +35,79 @@ public GitLabCollectionResponse<Group> GetAsync(GroupQuery query)
+        url = Utils.AddPageParams(url, page, perPage);
 
-        if (query.SkipGroups != null && query.SkipGroups.Any())
+        if (query is not null)
```
You can invert the condition and return early. That will remove one level of indentation

Toa741 commented on NGitLab/Impl/HttpRequestor.cs:
```diff
@@ -139,6 +173,14 @@ public virtual GitLabCollectionResponse<T> GetAllAsync<T>(string tailUrl)
         return new Enumerable<T>(_apiToken, GetAPIUrl(tailUrl), _options);
     }
 
+    private static string ReadText(Stream s) => new StreamReader(s).ReadToEnd();
```
StreamReader is disposable

belcher-rok committed:
Replace WebHeaderCollection with NameValueCollection

Do not use System.Net types in public APIs.

belcher-rok committed:
Revert all possible breaking changes

belcher-rok committed:
Update docs to use the 3rd person descriptive form

belcher-rok committed:
Replace `ToList` with `ToArray`

belcher-rok committed:
Keep static fields together

belcher-rok committed:
Dispose `StreamReader`

belcher-rok committed:
Return early in some cases to reduce indentation

belcher-rok committed:
Use `PagedResponse<T>` for all paged responses

belcher-rok committed:
Replace NameValueCollection with a IReadOnlyDictionary

belcher-rok committed:
Merge branch 'main' into main

belcher-rok committed:
Add unit tests for WebHeadersDictionaryAdaptor

belcher-rok committed:
Replace the custom IEnumerator impl with the IEnumerable's enumerator

belcher-rok commented:
> Hey @belcher-rok, thanks a lot for your contribution! This is great work!
> 
> Adding/improving comments was a nice touch too, but when documenting a method, could you please use the 3rd person (_descriptive form_, such as `gets`) rather than the 2nd person (_prescriptive form_, such as `get`)? This is a widespread standard which, unfortunately, we did not always comply with in the past.
> 
> Check out
> 
> * [The Google doc guidelines](https://developers.google.com/style/reference-verbs)
> * [An example from Microsoft doc](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.getattributes?view=net-6.0#definition)

@louis-z and @Toa741 Thanks for your reviews! You made the code better. I think I addressed all your suggestions except for one (see [above](https://github.com/ubisoft/NGitLab/pull/642#discussion_r1515294016)). Please let me know if you have any further requests.

A little background about me: My company is starting a project which utilizes a private GitLab instance. We are planning to use NGitLab as the client interface. It is quite likely that I will have additional PRs in the near future.

belcher-rok committed:
Merge branch 'main' into main

belcher-rok commented:
@louis-z and @Toa741 please let me know what I can do to help move this along.

louis-z commented:
> @louis-z and @Toa741 please let me know what I can do to help move this along.

Please address the remaining conversations and update your branch WRT main. Once it's done, this PR should get merged.

belcher-rok committed:
Use custom error message

belcher-rok commented:
> > @louis-z and @Toa741 please let me know what I can do to help move this along.
> 
> Please address the remaining conversations and update your branch WRT main. Once it's done, this PR should get merged.

All done.

belcher-rok committed:
Merge branch 'main' into main
